File: algorithms/lab00/python/task-3/sol1_stack.py
class MinStack:
    def __init__(self):
        # Stiva principala pentru stocarea elementelor
        self.st = []
        # Stiva auxiliara pentru stocarea elementelor minime
        self.minSt = []

    # Adauga elementul val in stiva.
    def push(self, val: int) -> None:
        self.st.append(val)

        # Daca stiva min este goala sau val este mai mic sau egal cu minimul curent,
        # il adauga in stiva min.
        if not self.minSt or val <= self.minSt[-1]:
            self.minSt.append(val)

    # Elimina elementul din varful stivei.
    def pop(self) -> None:
        if self.st[-1] == self.minSt[-1]:
            # Daca varful ambelor stive este acelasi, elimina si din stiva min.
            self.minSt.pop()
        self.st.pop()

    # Returneaza elementul din varful stivei.
    def top(self) -> int:
        return self.st[-1]

    # Returneaza elementul minim din stiva.
    def getMin(self) -> int:
        return self.minSt[-1]

File: algorithms/lab00/rust/task-3/sol1_stack.rs
// SPDX-License-Identifier: BSD-3-Clause

// Stiva principala pentru stocarea elementelor; stiva auxiliara pentru minime.
struct MinStack {
    st: Vec<i32>,
    min_st: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack {
            st: Vec::new(),
            min_st: Vec::new(),
        }
    }

    // Adauga elementul val in stiva.
    fn push(&mut self, val: i32) {
        self.st.push(val);
        if self.min_st.is_empty() || val <= *self.min_st.last().unwrap() {
            self.min_st.push(val);
        }
    }

    // Elimina elementul din varful stivei.
    fn pop(&mut self) {
        let top_val = *self.st.last().unwrap();
        if top_val == *self.min_st.last().unwrap() {
            self.min_st.pop();
        }
        self.st.pop();
    }

    // Returneaza elementul din varful stivei.
    fn top(&self) -> i32 {
        *self.st.last().unwrap()
    }

    // Returneaza elementul minim din stiva.
    fn get_min(&self) -> i32 {
        *self.min_st.last().unwrap()
    }
}
